File: AuthServer/AuthServer.API/Publishers/Services/Publisher.cs
using System.Text;
using AuthServer.API.Publisher.Interfaces;
using Newtonsoft.Json;
using RabbitMQ.Client;

namespace AuthServer.API.Publisher.Services;

public class Publisher : IPublisher
{
    private readonly IConnection _connection;
    public Publisher(
        IConnection connection
    )
    {
        _connection = connection;
    }
    public Task PublishAsync<T>(string exchange, string type, string routingKey, T data) where T : class
    {
        Console.WriteLine($"Publishing message on exchange {exchange} type {type} with routing key {routingKey}");
        // create exchange
        using var channel = _connection.CreateModel();
        channel.ExchangeDeclare(exchange, type);

        // serialize data
        var serializeData = JsonConvert.SerializeObject(data);
        var bytes = Encoding.UTF8.GetBytes(serializeData);
        // send data
        channel.BasicPublish(exchange, routingKey, false, null, bytes);
        return Task.CompletedTask;
    }
}

File: MovieServer/MovieServer.API/packages/MovieServer.Infrastructure/src/Broker/Publisher.ts
import { AmqpClient, IPublisher, Injectable } from "@movie/shared";
import { AMQPConnection } from "./AmqpClient";

@Injectable
export default class Publisher implements IPublisher {
    private readonly _amqpClient: AMQPConnection;

    constructor(
        @AmqpClient amqpClient: AMQPConnection
    ) {
        this._amqpClient = amqpClient;
    }

    async Publish<TRequest>(
        exchange: string,
        type: string,
        routingKey: string,
        data: TRequest,
        options?: Object): Promise<void> {
        try {
            // create channel
            const channel = await this._amqpClient.createChannel();
            // serialize message
            const serializeData = JSON.stringify(data);
            // create exchange
            const message = Buffer.from(serializeData, "utf-8");
            // publishing event
            await channel.assertExchange(exchange, type, { durable: false });
            channel.publish(exchange, routingKey, message, options);
            console.log(`Publish to exchange ${exchange} with routing key ${routingKey}`);
        } catch (err) {
            console.log(`error when publishing message ${err}`);
        }
    }

}